Events support

arjunguha opened the pull request:
Right now, if stopified code directly calls addEventListener, things may go very wrong. The problem is that when the browser invokes the event handler, it will start running just fine, but the first time the code suspends, it will crash. This is because stopified code must be invoked in the context of the main runtime loop (i.e., the loop in stopify-continuations/src/callcc/runtime/abstractRuntime.ts).

I suggest we don't support direct calls to addEventListener. Instead, this commit provides an API for external functions to use addEventListener. See clicks.html for a demo program. There are several new integration tests too.

arjunguha commented:
@jpolitz, @rachitnigam, and @baxtersa I'd like all of us to look over this.

arjunguha commented on stopify-continuations/src/runtime/abstractRuntime.ts:
```diff
@@ -171,15 +170,15 @@ export abstract class Runtime {
         else if(result.type === 'exception') {
           assert(this.mode,
             `execution completed in restore mode, error was: ${result.value}`);
-          throw result.value;
```
I don't believe there was any point in throwing or returning these values. Since the runtime system usually terminates on another turn (i.e., after yielding), the exception and return value would be lost. We could send the exception / result to the new onDone callback that I have defined.

arjunguha commented:
I should also point out that the contract for processEvents has an additional subtlety: the external function that calls `processEvent(callback)` needs to put all its event processing logic in `callback` since there is no guarantee that the event will be processed immediately. Moreover, it must make a *tail call* to stopified code and cannot rely on receiving a result from the stopified program.

Is this too lame? For functional event processing, do you need a result in Pyret?

jpolitz commented:
> Moreover, it must make a tail call to stopified code and cannot rely on receiving a result from the stopified program.

This sounds a lot like current constraints on code that interacts with Pyret. Pyret can only be called in tail position and only with a special helper that takes a callback.

Pyret uses state in JS to simulate getting a return value from reactors. It pauses the stack, then registers a bunch of event handlers, lets them run until the reactor is done, and keeps track of the current state when that's going on. When the reactor finishes, it restarts the paused thread with the reactor's current value.

I think `processEvent` needs to take another callback that receives the answer of calling event-processing function. A reactor would be something like:

```
function reactor(initValue, handlers) {
  return $__R.captureCC(function(k) {
    let currentValue = initValue;

    const wrapUpdateFun = (f) => (evt) => {
      runner.processEvent(() => f(evt), (answer) => {
        currentValue = answer;
        // NOTE(joe): need to somehow call stopWhen, which might be Stopified,
        // in a reasonably restart-y way. In Pyret we use safeCall.
        if(handlers.stopWhen(currentValue)) {
          finish();
        }
      };
    };
    const clickFun = wrapUpdateFun((evt) => handlers.onclick(currentValue, evt));
    const tickFun = wrapUpdateFun(() => handlers.ontick(currentValue));
    const finish = () {
      window.removeEventListener('click', clickFun);
      window.clearInterval(tickInterval);
      $__R.resumeFromSuspension(function() { return k(currentValue); });
    }
    window.addEventListener('click', clickFun);
    const tickInterval = window.setInterval(tickFun, 1000 / 24); // 24 fps
  });

}
```

arjunguha commented:
@jpolitz and others. Have a look at this. This is a processEvent API that should support reactors. The implementation details aren't exactly what we discussed. If there is way to make it simpler, that would be great.

jpolitz commented:
I really like naming this `turn`, because it makes it clear which browser/JS concept it's mapping to.

My thoughts:

- Now onYield returning true means "I want to be done with the current turn." I think this will interact in interesting ways with step debugging programs that register events. I think that the event-processing code could end up running even while the program is paused.
- Currently, the place where Pyret uses a custom onYield is to stop the whole program, which is a fine use of it with this semantics. The pausing semantics in the Stopify demo page I think would need to be rethought with this updated semantics

So I think as long as it's a property of Stopify that when onYield() returns true, it represents the end of a logical JavaScript turn, then this is good. Should a user expect that if their code runs in a single turn of JavaScript, onDone will only be called once with the result of that evaluation?

Another question: For Stopify's definition of a turn, can a user simulate blocking input in the same "turn" by this definition of endTurn?

baxtersa commented on stopify/src/runtime/suspend.ts:
```diff
@@ -43,7 +43,7 @@ export class RuntimeWithSuspend {
     // in a row will restart the computation.
     this.continuation = badResume;
-    return this.rts.resumeFromSuspension(cont);
+    return this.rts.resumeFromSuspension(cont, () => { });
```
In addition to what Joe is saying about stepping programs interacting weirdly with this, since we don't store the `onDone` callback this resumption isn't ever going to call it. This is only currently used by our web demo debugger (I think), but we may want to think about how to make events work in this context too.

arjunguha commented:
@baxtersa and @jpolitz: If I haven't broken anything, I've fixed events so that they work correctly with tests. I've also added an integration test that demonstrates that they work correctly.

@jpolitz you asked about "onYield returning true" interacting in strange ways with stepping. The onYield function that produces a boolean is an internal function and not one that is exposed to the user. I think it is a bit of a mess and that `RuntimeWithSuspend` and `AbstractRunner` should probably merge at some point. Have a look at my new integration -- that will hopefully convince you that events and pausing work well together.

@jpolitz you mentioned that Pyret set's a custom onYield function to pause the program. I think we need to figure out what the higher-level API for systems like Pyret should be. Updating an internal function that returns bool isn't so cool. Ideally, we'll figure out something similar to the AsyncRun API:

```
export interface AsyncRun {
  run(onDone: () => void,
    onYield?: () => void,
    onBreakpoint?: (line: number) => void): void;
  pause(onPaused: (line?: number) => void): void;
  resume(): void;
  setBreakpoints(line: number[]): void;
  step(onStep: (line: number) => void): void;
  pauseImmediate(callback: () => void): void;
  continueImmediate(result: any): void;
  processEvent(body: () => any, receiver: (x: Result) => void): void;
}
```

@jpolitz you asked if Stopify can simulate blocking input on the same turn (for Stopify's definition of turn). Oops! As defined, `pauseImmediate` captures the current continuation but not the event handler for the end of turn. We should probably fix that!

arjunguha commented:
There is probably a deeper design issue here: we cannot sleep at the top-level. Stopify has a continuation delimiter at each top-level statement.

arjunguha commented:
I added another test and fix. I've also added documentation. I am inclined to not fix the issue that Joe raises. I think a fix would require re-thinking delimit (which I'm happy to do). But, that is a much more significant change. I have a test case here that demonstrates Joe's issue:

 https://github.com/plasma-umass/Stopify/issues/350

The changes in this branch should have no effect on programs that don't use event handlers.

arjunguha commented:
Two more integration tests are needed:

1. The clicks.html demo needs to be turned into a test
2. We need a test where an event handler calls a function that uses pauseImmediate. E.g., an event handler that blocks for user input.

arjunguha commented:
@baxtersa if all these tests pass, I would like to make a release. Would you tell us how to post the documentation? In the `docs/conf.py` file, should I update both `version` and `release` to `0.2.0`?

baxtersa commented:
Update the numbers on `version` and `release`, and tag `master` with `0.2.0` once it lands.

The docs build at least two versions, `latest` tracking changes to `HEAD`, and `stable` tracking the most recent semantic versioning tag. I also archived the current version of the docs as `0.1.0`.
